Attachment cycling with onyx

ThatOneAwkwardGuy opened the pull request:
Implemented functionality to achieve the following statement:
"The attachments present in the chat should cycle when the user is viewing the modal and presses the left/right arrows on their keyboard"

### Details
Refactored code and moved Attachment modal so only one is created for ReportActionsView and one for the ReportActionCompose view. AttachmentModal now consumed visibleModal object from Onyx to know what to show and wether to be visible.

Also wrote getNextAttachment function which retrieves the next attachment using the current action and a boolean value indicating whether to move to the right or left of the current action.

### Fixed Issues
Implements #1038 

### Tests

1. Upload multiple attachments (Images and PDFs)
2. Once they have been uploaded, use the left and right arrow keys to move between all of them.
3. Go to the last attachment and confirm that moving to the right of the last attachment takes you to the first attachment in a cyclical manner.

1. Test that uploading attachments still work (Images, PDFs .etc)

### Tested On
- [x] Web
- [ ] Mobile Web
- [ ] Desktop
- [ ] iOS
- [ ] Android

### Screenshots

#### Web
<img width="2048" alt="Screenshot 2020-12-31 at 19 33 10" src="https://user-images.githubusercontent.com/13122879/103423562-25bda180-4b9f-11eb-9b68-22d2b26b6795.png">
<img width="2048" alt="Screenshot 2020-12-31 at 19 33 27" src="https://user-images.githubusercontent.com/13122879/103423566-2a825580-4b9f-11eb-9328-d3f53a567e82.png">

#### Mobile Web
no changes

#### Desktop
no changes

#### iOS
no changes

#### Android
no changes

ThatOneAwkwardGuy committed:
Working attachment cycling with onyx

marcaaron commented on src/components/RenderHTML.js:
```diff
@@ -113,23 +114,15 @@ function ImgRenderer({tnode}) {
-            sourceURL={source}
-            isAuthTokenRequired={isAttachment}
+        <TouchableOpacity
+            onPress={() => setAttachmentModalData({sourceURL: source, isAttachment, isModalOpen: true})}
```
I think this could be greatly improved if we restore the ability to pass props here. 

If we can't then for now, we might want to just stop using the `img` renderer for images. Instead we could parse the html attributes from the `message.html`. We would need some way to tell they are an attachment. Something similar to this:

https://github.com/Expensify/Expensify.cash/blob/b2c03fc5a8bd0d2c5a3f67847d132af1d41702e1/src/libs/actions/Report.js#L230-L237

But for all new actions we are adding here

https://github.com/Expensify/Expensify.cash/blob/b2c03fc5a8bd0d2c5a3f67847d132af1d41702e1/src/libs/actions/Report.js#L405-L412

I'd prefer to just wait until we can add props.

chiragsalian replied:
Yeah additionally it would also be amazing if we can add sequenceNumber when we call setAttachmentModalData just so that the attachmentModal knows where the current attachment belongs in the reportActions that looping everytime to find report action having the matching source url.

marcaaron commented on src/components/ComposeAttachmentModal.js:
```diff
@@ -0,0 +1,134 @@
+    session: {
+        key: ONYXKEYS.SESSION,
+    },
+})(AttachmentModal);
```
If we do make two separate components then we need to change the name we are using for this one.

marcaaron replied:
cc @tgolen since I know you are leaning towards two different components.

marcaaron commented on src/components/AttachmentModal.js:
```diff
@@ -128,6 +159,7 @@ class AttachmentModal extends Component {
 AttachmentModal.propTypes = propTypes;
 AttachmentModal.defaultProps = defaultProps;
 export default withOnyx({
+    visibleAttachment: {key: ONYXKEYS.ATTACHMENT_MODAL},
```
Ideally we would only need to store the source we want to show.

marcaaron commented on src/components/AttachmentModal.js:
```diff
@@ -74,52 +140,17 @@ class AttachmentModal extends Component {
-                        {this.state.sourceURL && (
-                            <AttachmentView sourceURL={sourceURL} file={this.state.file} />
+                        {this.props.visibleAttachment.sourceURL && (
+                            <AttachmentView sourceURL={sourceURL} file={this.props.visibleAttachment.file} />
```
A lot of these edits to change props to state can be avoided if we just pass the "active attachment" down from the parent instead of subscribing to updates here.

marcaaron commented on src/components/AttachmentModal.js:
```diff
@@ -42,28 +45,91 @@ const propTypes = {
+
+    // Function used to handle Left and Right key presses.
+    // This is a seperate function so that it can be subscribed/unsubscribed from the events
+    handleKeyPress = (event) => {
```
This method should be bound in the constructor or use an arrow function. We aren't using the class properties very much (if at all).

marcaaron commented on src/components/AttachmentModal.js:
```diff
@@ -42,28 +45,91 @@ const propTypes = {
+    // Function used to handle Left and Right key presses.
+    // This is a seperate function so that it can be subscribed/unsubscribed from the events
+    handleKeyPress = (event) => {
+        if (event.key === 'ArrowRight' || event.key === 'ArrowLeft') {
```
We have an existing library that we can use to bind these events called [`KeyboardShortcut`](https://github.com/Expensify/Expensify.cash/blob/master/src/libs/KeyboardShortcut/index.js)

marcaaron commented on src/components/AttachmentModal.js:
```diff
@@ -42,28 +45,91 @@ const propTypes = {
+        if (!prevProps.visibleAttachment.isModalOpen && isModalOpen) {
+            document.addEventListener('keydown', this.handleKeyPress, false);
+        } else {
+            document.removeEventListener('keydown', this.handleKeyPress, false);
```
I think `document` does not exist in native environments. So if you want to use it we'd need to have 2 versions of this component. `KeyboardShortcut` will solve this problem for you.

ThatOneAwkwardGuy replied:
KeyboardShortcuts unsubscribe wouldn't work for some reason, which was why I was forced to use document

marcaaron replied:
To clarify, `document` _cannot_ be used. It is breaking native platforms...

![2021-01-13_11-02-08](https://user-images.githubusercontent.com/32969087/104510137-ca23e880-558e-11eb-809c-e4d726c5b6f2.png)

marcaaron commented on src/components/AttachmentModal.js:
```diff
@@ -42,28 +45,91 @@ const propTypes = {
+    // A function to get the next attachment based on the current attachment being show in the report.
+    getNextAttachment = (toRight) => {
+        const {sortedReportActions, visibleAttachment: {sourceURL}} = this.props;
+        const attachments = sortedReportActions.filter(sortedReportAction => sortedReportAction.action.isAttachment);
```
We use the underscore variations of array methods for consistency.

marcaaron replied:
Using `isAttachment` is not reliable. If we want it to be we must set this when setting the report actions data here:

https://github.com/Expensify/Expensify.cash/blob/b2c03fc5a8bd0d2c5a3f67847d132af1d41702e1/src/libs/actions/Report.js#L405-L412

ThatOneAwkwardGuy replied:
I don't understand? I was using a property that was available to me via the action object.

marcaaron replied:
`isAttachment` is a property set for uploading attachments or attachments that are handled via Pusher. This does not come from the server therefore it is unreliable. e.g. here is a report action with an attachment after uploading, logging out, and logging back in.

![Screen Shot 2021-01-13 at 11 05 21 AM](https://user-images.githubusercontent.com/32969087/104510449-4f0f0200-558f-11eb-8014-0c98311ea0ed.png)

ThatOneAwkwardGuy replied:
ok

marcaaron commented on src/components/AttachmentModal.js:
```diff
@@ -42,28 +45,91 @@ const propTypes = {
+    }
+
+    // A function to get the next attachment based on the current attachment being show in the report.
+    getNextAttachment = (toRight) => {
```
NAB, I think this would be clearer if we used a constant to indicate which direction we are moving here and then pass `'right'` or `'left'`.

ThatOneAwkwardGuy replied:
But if we're not moving to the right, then the assumption is we're moving to the left.

ThatOneAwkwardGuy replied:
ok

marcaaron commented on src/components/AttachmentModal.js:
```diff
@@ -42,28 +45,91 @@ const propTypes = {
+        return imageGroup[1];
+    }
+
+    // A function to get the next attachment based on the current attachment being show in the report.
```
This comment should be a JSDoc

ThatOneAwkwardGuy replied:
Ok

marcaaron commented on src/components/AttachmentModal.js:
```diff
@@ -42,28 +45,91 @@ const propTypes = {
+        const currentAttachment = attachments.find((attachment) => {
+            const html = lodashGet(attachment, ['action', 'message', 0, 'html'], '');
+            return html.includes(sourceURL);
+        });
```
NAB, What happens if we do not find it? It seems not possible but maybe we should return early or log something?

ThatOneAwkwardGuy replied:
Ok

marcaaron commented on src/components/AttachmentModal.js:
```diff
@@ -42,28 +45,91 @@ const propTypes = {
+        const attachmentSource = this.getAttachmentSource(html);
+        setAttachmentModalData({
+            sourceURL: attachmentSource, isAttachment: true, isModalOpen: true, file: null,
+        });
```
I think we could simplify this process by just creating an array of actions _once_ (until the actions update again) and pass them to the modal as a prop instead of all actions. Then we can just find the index of our action in that small array. Look left and right +1/-1 and we've found the right action if it's out of bounds then return the attachment at `0` or `attachmentArray.length - 1`.

ThatOneAwkwardGuy replied:
the next action may not contain an attachment.

marcaaron replied:
Right, I am suggesting we _only_ pass actions with attachments and not all actions.

ThatOneAwkwardGuy replied:
ok

marcaaron reviewed (changes requested):
Moving in a better direction thanks! There are some problems still. If we can succeed in parsing the html instead of using the `RenderHTML` component we should be able to improve this.

marcaaron commented:
also adding @tgolen

ThatOneAwkwardGuy commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -242,14 +246,17 @@ class ReportActionsView extends React.Component {
-                initialRowHeight={32}
-            />
+            <>
+                <AttachmentModal sortedReportActions={this.sortedReportActions} />
```
Potentially, I wanted to have most of the changes in the AttachmentModal file which is why I simply passed the actions straight into the modal. If you want to filter it before and then pass it in, that's and easy chage.

marcaaron replied:
>I wanted to have most of the changes in the AttachmentModal file which is why I simply passed the actions straight into the modal

Cool, I think it should be done differently. The `AttachmentModal` should be concerned with attachments. It should not be concerned with all report actions or be responsible for filtering them and extracting the attachments inside them.

cc @chiragsalian just to make sure we are on the same page with this.

chiragsalian replied:
Actually yeah i definitely agree with this. If we filter all attachments here and pass just `attachmentActions` it would be so much lighter on the logic in `AttachmentModal` since it won't have to filter through all report actions on every keypress.

ThatOneAwkwardGuy replied:
ok

ThatOneAwkwardGuy commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -54,6 +57,7 @@ class ReportActionsView extends React.Component {
     }
 
     componentDidMount() {
+        clearAttachmentModalData();
```
The AttachmentModal never actually unmounts. It only ever has its isModalOpen property toggle.

ThatOneAwkwardGuy commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -209,7 +210,7 @@ class ReportActionCompose extends React.Component {
                             </>
                         )}
-                    </AttachmentModal>
+                    </ComposeAttachmentModal>
```
Here:
https://github.com/Expensify/Expensify.cash/pull/1149#pullrequestreview-564499774

tgolen suggests having separate components to keep the code clean, therefore that's why I've changed the component to its own separate  ComposeAttachmentModal.

marcaaron replied:
@tgolen That seems ok to me, but the consequence is that we have shared logic being duplicated.

ThatOneAwkwardGuy replied:
At this point I would say the only thing they share is that they instantiate a modal, their purposes are quite different now and will likely get more different in the future if you decide to add upload progress .etc to the compose modal.

I believe writing a singular modal to fulfil both use cases would be more confusing.

ThatOneAwkwardGuy commented on src/libs/actions/Report.js:
```diff
@@ -649,6 +650,33 @@ function handleReportChanged(report) {
+        sourceURL, isAttachment, isModalOpen, file,
+    });
+}
+
```
Doesn't this only occur when the file is uploaded, therefore would only be shown in the ComposeAttachmentModal component?

ThatOneAwkwardGuy replied:
Also we're storing the state like this because the AttachmentModal's isModalOpen property needs to be controlled from a centralised position.

marcaaron replied:
Ok, so we need the `sourceURL`, `isAttachment`, and `isModalOpen`? Do we need `file`?

ThatOneAwkwardGuy replied:
I have no clue what the initial purpose of what the file attribute was in the codebase, I assume it was to show the file name for PDFs or something. But I simply kept it because I didn't think it was my place to be deleting functionality

ThatOneAwkwardGuy commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -1,19 +1,20 @@
+import paperClipIcon from '../../../../assets/images/icon-paper-clip.png';
+import sendIcon from '../../../../assets/images/icon-send.png';
+import styles from '../../../styles/styles';
+import themeColors from '../../../styles/themes/default';
```
Ok

ThatOneAwkwardGuy committed:
Added JSDoc for getNextAttachment and fixed import changes

ThatOneAwkwardGuy committed:
Filter change and modal listener fix

- Changed filter to use underscore
- Fixed modal listener logic where listener would stop workin after first trigger.

chiragsalian commented on src/components/AttachmentModal.js:
```diff
@@ -42,28 +45,98 @@ const propTypes = {
+        const actionsToSearch = toRight
+            ? actions.slice(sequenceNumber)
+            : actions.slice(0, sequenceNumber !== 0 ? sequenceNumber - 1 : 0).reverse();
+        let nextAttachment = actionsToSearch.find(actionResult => actionResult.action.isAttachment);
```
This logic here seems pretty complicated. I think you can achieve the same results on this method using just a few lines of the underscore [chain command](https://underscorejs.org/#chain). So basically the whole logic in this method could be simplified to something like this,
```jsx
const currentSequenceNumber = this.props.visibleAttachment.sequenceNumber;
const nextAttachmentAction = _.chain(this.props.sortedReportActions)
    .pluck('action')
    .filter(action => (toRight ? action.sequenceNumber > currentSequenceNumber : action.sequenceNumber < currentSequenceNumber))
    .find(actions => actions.isAttachment)
    .value();
```

ThatOneAwkwardGuy replied:
As I mentioned above:

`I read the requirement saying "The attachments present in the chat should cycle" literally. If there is a consensus that its overkill, I can remove it.`

jboniface replied:
My apologies, I talked this over with @chiragsalian and I think he makes a good point that the carousel (i.e. cycling) behavior is unnecessary in this context, as there is unlikely to be a context where you need to loop back to the first image you ever uploaded. i think we can probably cut this behavior.

chiragsalian replied:
Well the example i posted is just a way to clean up the logic to make it easier to read. Like i mentioned, in untested which also means you can tweak it as needed. The idea is using chain makes it read. For example if carousel was needed the logic would be something like this, (untested)
```jsx
const nextReportActions = // slice reports actions in the correct direction to the end or beginning based on direction.
const carouselReportActions = // these are the reports actions that loop and we can slice on the opposing direction and reverse the list.

// Then we'd use chain again to do
const nextAttachmentAction = _.chain(nextReportActions)
    .union(carouselReportActions)
    .pluck('action')
    .filter(action => (toRight ? action.sequenceNumber > currentSequenceNumber : action.sequenceNumber < currentSequenceNumber))
    .find(actions => actions.isAttachment)
    .value();
```

But anyway from this [comment](https://github.com/Expensify/Expensify.cash/issues/1038#issuecomment-759716134) it looks like the carousel is not needed so can you still clean up the logic using chain.

ThatOneAwkwardGuy replied:
ok

chiragsalian commented on src/components/AttachmentModal.js:
```diff
@@ -42,28 +45,98 @@ const propTypes = {
+    }
+
+    // Function used to handle Left and Right key presses.
+    // This is a seperate function so that it can be subscribed/unsubscribed from the events
```
incorrect JSDoc style

ThatOneAwkwardGuy replied:
ok

chiragsalian commented on src/libs/actions/Report.js:
```diff
@@ -649,6 +650,33 @@ function handleReportChanged(report) {
+ * @param {Boolean} reportID.isAttachment
+ * @param {Boolean} reportID.isModalOpen
+ * @param {Object} reportID.file
+ * @param {String} reportID.file.name
```
unused param

ThatOneAwkwardGuy replied:
Im confused, file defaults to null, but it can also be an object with a name attribute

chiragsalian replied:
True its weird that file defaults to null. But even either way `reportID.file` is Object so its param type is correct. We don't mention child params in JSDocs unless its' explicitly used by the method and since this method doesn't directly use `reportID.file.name` we can remove its mention here. Plus some editors will even complain about it,
![image](https://user-images.githubusercontent.com/14356145/104511108-19731480-55a1-11eb-92e1-b8c751592c89.png)

ThatOneAwkwardGuy replied:
ok

ThatOneAwkwardGuy commented on src/ONYXKEYS.js:
```diff
@@ -47,4 +47,6 @@ export default {
         REPORT_USER_IS_TYPING: 'reportUserIsTyping_',
     },
+
+    ATTACHMENT_MODAL: 'attachmentModalData',
```
ok

ThatOneAwkwardGuy commented on src/components/AttachmentModal.js:
```diff
@@ -1,39 +1,42 @@
+    // Current visible attachment
+    visibleAttachment: PropTypes.shape({
+        sourceURL: PropTypes.string,
+        isAttachment: PropTypes.bool,
```
Honestly, no clue, I just saw that isAttachment can be either true or false in the codebase and that it's used on appending an auth code to get the attachment data so I didn't want to miss it out.

ThatOneAwkwardGuy commented on src/components/AttachmentModal.js:
```diff
@@ -42,28 +45,98 @@ const propTypes = {
+        }
+    }
+
+    // Using regex to get the src attribute from the html img tag
```
ok

ThatOneAwkwardGuy commented on src/components/AttachmentModal.js:
```diff
@@ -42,28 +45,98 @@ const propTypes = {
+        }
+
+        // Finding the current attachment inside the report using the sourceURL
+        const currentAttachment = attachments.find((attachment) => {
```
Not currently possible as ImgRenderer does not pass any of that information into the TouchableOpacity. So when that's pressed, there would be no sequence number. I can write code that sets the sequence number once left or right arrow has been pressed the first time, however, that means there's two different forms of logic that would need to be written, one for when no sequence number is available and one for when it is.

chiragsalian replied:
Ah true i just noticed that. I see marc discussed something similar [here](https://github.com/Expensify/Expensify.cash/pull/1210#discussion_r555441094) so I'll resolve this and we can continue there as needed.

ThatOneAwkwardGuy commented on src/components/AttachmentModal.js:
```diff
@@ -42,28 +45,98 @@ const propTypes = {
+
+        // If there is no next attachment in the left or right direction,
+        // then we need to loop to get the next attachment after the end of the array
+        if (!nextAttachment) {
```
I read the requirement saying "The attachments present in the chat should cycle" literally. If there is a consensus that its overkill, I can remove it.

chiragsalian replied:
Ah okay cool. I missed reading that requirement. All good here 👍

chiragsalian replied:
As per this [comment](https://github.com/Expensify/Expensify.cash/issues/1038#issuecomment-759716134) let's remove this block.

jboniface replied:
> My apologies, I talked this over with @chiragsalian and I think he makes a good point that the carousel (i.e. cycling) behavior is unnecessary in this context, as there is unlikely to be a context where you need to loop back to the first image you ever uploaded. i think we can probably cut this behavior.

ThatOneAwkwardGuy replied:
ok

ThatOneAwkwardGuy committed:
JSDoc fixes, getNextAttachment changes

changes getNextAttachment argument from toRight to direction.

Removed object destructuring for readability.

ThatOneAwkwardGuy commented on src/components/AttachmentModal.js:
```diff
@@ -42,28 +45,98 @@ const propTypes = {
     }
 
+    componentDidUpdate(prevProps) {
+        const {visibleAttachment: {isModalOpen}} = this.props;
```
ok

ThatOneAwkwardGuy committed:
Review Changes

- KeyboardShortcut usages instead of document.

- AttachmentModal only now receives attachments

- Removed reportID.file.name from JSDoc definition

chiragsalian commented:
Closing PR. Marc and i discussed about this issue for a while and we feel like we should close this PR for now. We'll be splitting the original GH issue into multiple smaller issues to discuss and improve renderHTML and cycling attachments and once we've settled with a technical plan we'll be posting those in upwork.

Thank you for your time and work. Closing PR now.
